This notebook will be mainly used for the Capstone Project in Coursera Course 9 of IBM Data Science Professional Certificate.




In [188]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import numpy as np

In [189]:
print("Hello Capstone Project Course!")

Hello Capstone Project Course!


In [190]:
# url of the web page we target for scraping
# https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
# in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe 

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood.
Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.
4. Submit a link to your Notebook on your Github repository. (10 marks)

Note: There are different website scraping libraries and packages in Python. One of the most common packages is BeautifulSoup. Here is the package's main documentation page: http://beautiful-soup-4.readthedocs.io/en/latest/
The package is so popular that there is a plethora of tutorials and examples of how to use it. Here is a very good Youtube video on how to use the BeautifulSoup package: https://www.youtube.com/watch?v=ng2o98k983k
Use the BeautifulSoup package or any other way you are comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe


===== While researching on how using BeautifulSoup4 and lxml parser - see 2 cells below put in comments - we discovered the read_html function in pandas that will work fine for the first part of the work to be done on our data frame. =====

In [191]:
import requests
website_url = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").text

#from bs4 import BeautifulSoup
#soup = BeautifulSoup(website_url, 'lxml')
#print(soup.prettify())

In [192]:
# Under table class ‘wikitable sortable’ we have links with postal codes and boroughs names as title.
#My_table = soup.find('table',{'class':'wikitable sortable'})
#My_table

In [193]:
raw_df_postcode = pd.read_html(website_url, header=0, attrs={"class":"wikitable sortable"})[0]
raw_df_postcode


Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M5A  Downtown Toronto   
6        M6A        North York   
7        M6A        North York   
8        M7A      Queen's Park   
9        M8A      Not assigned   
10       M9A         Etobicoke   
11       M1B       Scarborough   
12       M1B       Scarborough   
13       M2B      Not assigned   
14       M3B        North York   
15       M4B         East York   
16       M4B         East York   
17       M5B  Downtown Toronto   
18       M5B  Downtown Toronto   
19       M6B        North York   
20       M7B      Not assigned   
21       M8B      Not assigned   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M9B         Etobicoke   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M1C       Scarborough   
30       M2C      Not assigned   
31       M3C        North York   
32       M3C        North York   
33       M4C         East York   
34       M5C  Downtown Toronto   
35       M6C              York   
36       M7C      Not assigned   
37       M8C      Not assigned   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M9C         Etobicoke   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M1E       Scarborough   
45       M2E      Not assigned   
46       M3E      Not assigned   
47       M4E      East Toronto   
48       M5E  Downtown Toronto   
49       M6E              York   
50       M7E      Not assigned   
51       M8E      Not assigned   
52       M9E      Not assigned   
53       M1G       Scarborough   
54       M2G      Not assigned   
55       M3G      Not assigned   
56       M4G         East York   
57       M5G  Downtown Toronto   
58       M6G  Downtown Toronto   
59       M7G      Not assigned   
60       M8G      Not assigned   
61       M9G      Not assigned   
62       M1H       Scarborough   
63       M2H        North York   
64       M3H        North York   
65       M3H        North York   
66       M3H        North York   
67       M4H         East York   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M5H  Downtown Toronto   
71       M6H      West Toronto   
72       M6H      West Toronto   
73       M7H      Not assigned   
74       M8H      Not assigned   
75       M9H      Not assigned   
76       M1J       Scarborough   
77       M2J        North York   
78       M2J        North York   
79       M2J        North York   
80       M3J        North York   
81       M3J        North York   
82       M4J         East York   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M5J  Downtown Toronto   
86       M6J      West Toronto   
87       M6J      West Toronto   
88       M7J      Not assigned   
89       M8J      Not assigned   
90       M9J      Not assigned   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M1K       Scarborough   
94       M2K        North York   
95       M3K        North York   
96       M3K        North York   
97       M4K      East Toronto   
98       M4K      East Toronto   
99       M5K  Downtown Toronto   
100      M5K  Downtown Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M6K      West Toronto   
104      M7K      Not assigned   
105      M8K      Not assigned   
106      M9K      Not assigned   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M1L       Scarborough   
110      M2L        North York   
111      M2L        North York   
112      M3L        North York   
113      M4L      East Toronto   
114      M4L      East Toronto   
115      M5L  Downtown Toronto   
116      M5L  Downtown Tor

In [194]:
# clean the postcode table:
# Adjust the columns names as requested "PostalCode", "Borough", and "Neighborhood"
PostalCodeTable = raw_df_postcode.rename(columns={"Postcode":"PostalCode", "Borough":"Borough", "Neighbourhood":"Neighborhood"})
#PostalCodeTable.tail()

# Drop the rows with Borough == 'Not assigned'
PostalCodeTable=PostalCodeTable.drop(PostalCodeTable.index[PostalCodeTable['Borough']=='Not assigned'])
PostalCodeTable


PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M5A  Downtown Toronto   
6          M6A        North York   
7          M6A        North York   
8          M7A      Queen's Park   
10         M9A         Etobicoke   
11         M1B       Scarborough   
12         M1B       Scarborough   
14         M3B        North York   
15         M4B         East York   
16         M4B         East York   
17         M5B  Downtown Toronto   
18         M5B  Downtown Toronto   
19         M6B        North York   
22         M9B         Etobicoke   
23         M9B         Etobicoke   
24         M9B         Etobicoke   
25         M9B         Etobicoke   
26         M9B         Etobicoke   
27         M1C       Scarborough   
28         M1C       Scarborough   
29         M1C       Scarborough   
31         M3C        North York   
32         M3C        North York   
33         M4C         East York   
34         M5C  Downtown Toronto   
35         M6C              York   
38         M9C         Etobicoke   
39         M9C         Etobicoke   
40         M9C         Etobicoke   
41         M9C         Etobicoke   
42         M1E       Scarborough   
43         M1E       Scarborough   
44         M1E       Scarborough   
47         M4E      East Toronto   
48         M5E  Downtown Toronto   
49         M6E              York   
53         M1G       Scarborough   
56         M4G         East York   
57         M5G  Downtown Toronto   
58         M6G  Downtown Toronto   
62         M1H       Scarborough   
63         M2H        North York   
64         M3H        North York   
65         M3H        North York   
66         M3H        North York   
67         M4H         East York   
68         M5H  Downtown Toronto   
69         M5H  Downtown Toronto   
70         M5H  Downtown Toronto   
71         M6H      West Toronto   
72         M6H      West Toronto   
76         M1J       Scarborough   
77         M2J        North York   
78         M2J        North York   
79         M2J        North York   
80         M3J        North York   
81         M3J        North York   
82         M4J         East York   
83         M5J  Downtown Toronto   
84         M5J  Downtown Toronto   
85         M5J  Downtown Toronto   
86         M6J      West Toronto   
87         M6J      West Toronto   
91         M1K       Scarborough   
92         M1K       Scarborough   
93         M1K       Scarborough   
94         M2K        North York   
95         M3K        North York   
96         M3K        North York   
97         M4K      East Toronto   
98         M4K      East Toronto   
99         M5K  Downtown Toronto   
100        M5K  Downtown Toronto   
101        M6K      West Toronto   
102        M6K      West Toronto   
103        M6K      West Toronto   
107        M1L       Scarborough   
108        M1L       Scarborough   
109        M1L       Scarborough   
110        M2L        North York   
111        M2L        North York   
112        M3L        North York   
113        M4L      East Toronto   
114        M4L      East Toronto   
115        M5L  Downtown Toronto   
116        M5L  Downtown Toronto   
117        M6L        North York   
118        M6L        North York   
119        M6L        North York   
122        M9L        North York   
123        M1M       Scarborough   
124        M1M       Scarborough   
125        M1M       Scarborough   
126        M2M        North York   
127        M2M        North York   
128        M3M        North York   
129        M4M      East Toronto   
130        M5M        North York   
131        M5M        North York   
132        M6M              York   
133        M6M              York   
134        M6M              York   
135        M6M              York   
138        M9M        North York   
139        M9M        North York   
140        M1N       Scarborough   
141        M1N       Scarborough   
142     

In [195]:
# continue cleaning
# When a row shows Neighborhood == Not assigned, then Neighborhood gets the value of Borough
def NotAssign (a,b): return a if b=="Not assigned" else b
for ind in PostalCodeTable.index.values: PostalCodeTable['Neighborhood'][ind]=NotAssign(PostalCodeTable['Borough'][ind],PostalCodeTable['Neighborhood'][ind])
PostalCodeTable

PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M5A  Downtown Toronto   
6          M6A        North York   
7          M6A        North York   
8          M7A      Queen's Park   
10         M9A         Etobicoke   
11         M1B       Scarborough   
12         M1B       Scarborough   
14         M3B        North York   
15         M4B         East York   
16         M4B         East York   
17         M5B  Downtown Toronto   
18         M5B  Downtown Toronto   
19         M6B        North York   
22         M9B         Etobicoke   
23         M9B         Etobicoke   
24         M9B         Etobicoke   
25         M9B         Etobicoke   
26         M9B         Etobicoke   
27         M1C       Scarborough   
28         M1C       Scarborough   
29         M1C       Scarborough   
31         M3C        North York   
32         M3C        North York   
33         M4C         East York   
34         M5C  Downtown Toronto   
35         M6C              York   
38         M9C         Etobicoke   
39         M9C         Etobicoke   
40         M9C         Etobicoke   
41         M9C         Etobicoke   
42         M1E       Scarborough   
43         M1E       Scarborough   
44         M1E       Scarborough   
47         M4E      East Toronto   
48         M5E  Downtown Toronto   
49         M6E              York   
53         M1G       Scarborough   
56         M4G         East York   
57         M5G  Downtown Toronto   
58         M6G  Downtown Toronto   
62         M1H       Scarborough   
63         M2H        North York   
64         M3H        North York   
65         M3H        North York   
66         M3H        North York   
67         M4H         East York   
68         M5H  Downtown Toronto   
69         M5H  Downtown Toronto   
70         M5H  Downtown Toronto   
71         M6H      West Toronto   
72         M6H      West Toronto   
76         M1J       Scarborough   
77         M2J        North York   
78         M2J        North York   
79         M2J        North York   
80         M3J        North York   
81         M3J        North York   
82         M4J         East York   
83         M5J  Downtown Toronto   
84         M5J  Downtown Toronto   
85         M5J  Downtown Toronto   
86         M6J      West Toronto   
87         M6J      West Toronto   
91         M1K       Scarborough   
92         M1K       Scarborough   
93         M1K       Scarborough   
94         M2K        North York   
95         M3K        North York   
96         M3K        North York   
97         M4K      East Toronto   
98         M4K      East Toronto   
99         M5K  Downtown Toronto   
100        M5K  Downtown Toronto   
101        M6K      West Toronto   
102        M6K      West Toronto   
103        M6K      West Toronto   
107        M1L       Scarborough   
108        M1L       Scarborough   
109        M1L       Scarborough   
110        M2L        North York   
111        M2L        North York   
112        M3L        North York   
113        M4L      East Toronto   
114        M4L      East Toronto   
115        M5L  Downtown Toronto   
116        M5L  Downtown Toronto   
117        M6L        North York   
118        M6L        North York   
119        M6L        North York   
122        M9L        North York   
123        M1M       Scarborough   
124        M1M       Scarborough   
125        M1M       Scarborough   
126        M2M        North York   
127        M2M        North York   
128        M3M        North York   
129        M4M      East Toronto   
130        M5M        North York   
131        M5M        North York   
132        M6M              York   
133        M6M              York   
134        M6M              York   
135        M6M              York   
138        M9M        North York   
139        M9M        North York   
140        M1N       Scarborough   
141        M1N       Scarborough   
142     

In [196]:
# as requested : append neighborhoods with the same postal code within the same neighborhood cell, separated by a comma
TableGroup=PostalCodeTable.groupby(['PostalCode','Borough'], sort=False, as_index=False).agg(lambda x: ', '.join(set(x)))

In [197]:
TableGroup

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A      Queen's Park   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
3                     Lawrence 

In [198]:
TableGroup.shape

(103, 3)

# here starts cells for the 2nd submit of week3, thanks.

In [199]:
!conda install -c conda-forge geocoder --yes

Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /opt/conda/envs/DSX-Python35:
#
geocoder                  1.38.1                     py_0    conda-forge


In [200]:
import geocoder

In [201]:
# First, let's try geocoder just for one postal code
# initialize your variable to None
#lat_lng_coords = None
#postal_code = "M3A"

# loop until you get the coordinates
#while(lat_lng_coords is None):
#    g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
#lat_lng_coords = g.latlng
#lat_lng_coords

#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

####### as no answer received from these requests let's focus on analysis of the data and choose the other option wchich is 
####### getting data from the .csv provided

In [202]:
Geo_Coord_Table=pd.read_csv("https://cocl.us/Geospatial_data")
Geo_Coord_Table.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [203]:
# check we have the same number of postal codes 
Geo_Coord_Table.shape

(103, 3)

In [204]:
Geo_Coord_Table.rename(columns={"Postal Code":"PostalCode"}, inplace=True)

In [205]:
Geo_Coord_Table.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [206]:
# Let sort both dataframes by postal code column.
Geo_Coord_Table.sort_values(by = ['PostalCode'], inplace=True)
TableGroup.sort_values(by = ['PostalCode'], inplace=True)

In [207]:
Toronto_Neighb_LatLon = TableGroup.merge(Geo_Coord_Table, left_on='PostalCode', right_on='PostalCode')

In [208]:
Toronto_Neighb_LatLon.head(20)

PostalCode      Borough                                       Neighborhood  \
0         M1B  Scarborough                                     Malvern, Rouge   
1         M1C  Scarborough             Port Union, Rouge Hill, Highland Creek   
2         M1E  Scarborough                  Morningside, Guildwood, West Hill   
3         M1G  Scarborough                                             Woburn   
4         M1H  Scarborough                                          Cedarbrae   
5         M1J  Scarborough                                Scarborough Village   
6         M1K  Scarborough        Ionview, Kennedy Park, East Birchmount Park   
7         M1L  Scarborough                    Golden Mile, Oakridge, Clairlea   
8         M1M  Scarborough    Scarborough Village West, Cliffcrest, Cliffside   
9         M1N  Scarborough                        Birch Cliff, Cliffside West   
10        M1P  Scarborough  Dorset Park, Wexford Heights, Scarborough Town...   
11        M1R  Scarborough                                  Wexford, Maryvale   
12        M1S  Scarborough                                          Agincourt   
13        M1T  Scarborough            Clarks Corners, Tam O'Shanter, Sullivan   
14        M1V  Scarborough  L'Amoreaux East, Agincourt North, Steeles East...   
15        M1W  Scarborough                      Steeles West, L'Amoreaux West   
16        M1X  Scarborough                                        Upper Rouge   
17        M2H   North York                                  Hillcrest Village   
18        M2J   North York                       Oriole, Fairview, Henry Farm   
19        M2K   North York                                    Bayview Village   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577  
15  43.799525 -79.318389  
16  43.836125 -79.205636  
17  43.803762 -79.363452  
18  43.778517 -79.346556  
19  43.786947 -79.385975

In [209]:
Toronto_Neighb_LatLon.shape

(103, 5)

# here starts cells for the 3rd submit of week3, thanks.

In [210]:
import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /opt/conda/envs/DSX-Python35:
#
folium                    0.5.0                      py_0    conda-forge
Libraries imported.


In [211]:
# create a map of Toronto area with neighborhoods superimposed on top
# Let's center on Downtwon Toronto - The first postal code in the list for Dtwn Toronto is M4W (row 50)
latitude = Toronto_Neighb_LatLon.loc[Toronto_Neighb_LatLon['PostalCode']=='M4W',['Latitude']]
latitude
lat=latitude.values[0,0]
longitude = Toronto_Neighb_LatLon.loc[Toronto_Neighb_LatLon['PostalCode']=='M4W',['Longitude']]
longitude
lon=longitude.values[0,0]
#longitude = location.longitude
print('The geograpical coordinate of DTWN Toronto are {}, {}.'.format(lat, lon))

The geograpical coordinate of DTWN Toronto are 43.6795626, -79.37752940000001.


In [212]:
TOR_NEIGHB = Toronto_Neighb_LatLon

# create map of Toronto area with latitude and longitude values
map_toronto = folium.Map(location=[lat, lon], zoom_start=10)

# add markers to map
for lati, long, postcode, borough in zip(TOR_NEIGHB['Latitude'], TOR_NEIGHB['Longitude'], TOR_NEIGHB['PostalCode'], TOR_NEIGHB['Borough']):
    label = '{}, {}'.format(postcode, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lati, long],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Now let's explore and cluster our Toronto neighborhoods. We will use the table below and work with postal codes as we did above for the map and markers.

In [213]:
TOR_NEIGHB.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Port Union, Rouge Hill, Highland Creek  43.784535   
2        M1E  Scarborough       Morningside, Guildwood, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [214]:
TOR_NEIGHB.tail()

PostalCode    Borough                                       Neighborhood  \
98         M9N       York                                             Weston   
99         M9P  Etobicoke                                          Westmount   
100        M9R  Etobicoke  Martin Grove Gardens, Kingsview Village, Richv...   
101        M9V  Etobicoke  Silverstone, Beaumond Heights, South Steeles, ...   
102        M9W  Etobicoke                                          Northwest   

      Latitude  Longitude  
98   43.706876 -79.518188  
99   43.696319 -79.532242  
100  43.688905 -79.554724  
101  43.739416 -79.588437  
102  43.706748 -79.594054

Define Foursquare Credentials and Version

In [215]:
TOR_NEIGHB.shape

(103, 5)

In [216]:
CLIENT_ID = 'AU41FQ2PT2Q5G2CD2H3VFHKKBHT4C211LDKE0HIQMFRWT0QE' # your Foursquare ID
CLIENT_SECRET = 'J2GZUGPCFEBCR43E2KMM02QPD3ZB50LK4ALSTWU33SFHUVIA' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: AU41FQ2PT2Q5G2CD2H3VFHKKBHT4C211LDKE0HIQMFRWT0QE
CLIENT_SECRET:J2GZUGPCFEBCR43E2KMM02QPD3ZB50LK4ALSTWU33SFHUVIA


Let's explore the first area in our database. This is postal code M1B in the borough Scarborough. It is located in the eastern part of our marked areas on the map. As several neighborhoods can have the same postal code, we will identify areas with their postal code. We can always refer to the table to see in which borough this postal code falls, and to the map to see which marker this area is.
We have 103 different areas to explore, corresponding to the 103 postal codes we have in our table. For each postal code, we have a corresponding (latitude, longitude) that we use to plot a marker on the map and we use to explore neigborhoods with foursquare data.

In [217]:
CP_lat = TOR_NEIGHB.loc[0, 'Latitude'] # postal code latitude value
CP_lon = TOR_NEIGHB.loc[0, 'Longitude'] # postal code longitude value

CP_name = TOR_NEIGHB.loc[0, 'PostalCode'] # Postal Code name
Borough_name = TOR_NEIGHB.loc[0, 'Borough'] # this Postal Code belongs to this Borough

print('Within the borough {}, latitude and longitude values of postal code {} are {}, {}.'.format(Borough_name,
                                                                                                  CP_name,
                                                                                                  CP_lat,
                                                                                                  CP_lon))

Within the borough Scarborough, latitude and longitude values of postal code M1B are 43.806686299999996, -79.19435340000001.


#### Now, let's get the top 100 venues that are in postal code M1B in Scarborough within a radius of 500 meters.
First, let's create the GET request URL. Name your URL **url**.

In [218]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    CP_lat, 
    CP_lon, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=AU41FQ2PT2Q5G2CD2H3VFHKKBHT4C211LDKE0HIQMFRWT0QE&client_secret=J2GZUGPCFEBCR43E2KMM02QPD3ZB50LK4ALSTWU33SFHUVIA&v=20180605&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

Send the GET request and examine the results

In [219]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c5402624434b9740e892435'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4bb6b9446edc76b0d771311c-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/fastfood_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d16e941735',
         'name': 'Fast Food Restaurant',
         'pluralName': 'Fast Food Restaurants',
         'primary': True,
         'shortName': 'Fast Food'}],
       'id': '4bb6b9446edc76b0d771311c',
       'location': {'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'Morningside & Sheppard',
        'distance': 387,
        'formattedAddress': ['Toronto ON', 'Canada'],
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.80744841934756,
          'ln

We see that with a radius 500 there is only 1 result, which surely corresponds to the area we are currently exploring.
Let's see how a greater radius will work

In [220]:
# now with radius = 1000
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 1000 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    CP_lat, 
    CP_lon, 
    radius, 
    LIMIT)
url # display URL

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c5402621ed2193b434684fa'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4d669cba83865481c948fa53-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/shops/spa_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d1ed941735',
         'name': 'Spa',
         'pluralName': 'Spas',
         'primary': True,
         'shortName': 'Spa'}],
       'id': '4d669cba83865481c948fa53',
       'location': {'address': '8130 Sheppard Ave E',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'Morningside Ave',
        'distance': 595,
        'formattedAddress': ['8130 Sheppard Ave E (Morningside Ave)',
         'Toronto ON M1B 3W3',
         'Canada'],
        'labeledLatLngs': [{'label': 'd

With a radius of 1000 instead of 500 we get 16 results.
We may think of keeping a 1000 radius to get more results but we would have to keep the same radius value for each area to explore of course, in order to keep the data consistent. And 500 is more relevant than 1000 for many areas closer to downtown Toronto.
Let's double check with trying with areas within downtown Toronto borough!

In [221]:
#TOR_NEIGHB
TOR_NEIGHB.where(TOR_NEIGHB['Borough']=='Downtown Toronto').dropna()

PostalCode           Borough  \
50        M4W  Downtown Toronto   
51        M4X  Downtown Toronto   
52        M4Y  Downtown Toronto   
53        M5A  Downtown Toronto   
54        M5B  Downtown Toronto   
55        M5C  Downtown Toronto   
56        M5E  Downtown Toronto   
57        M5G  Downtown Toronto   
58        M5H  Downtown Toronto   
59        M5J  Downtown Toronto   
60        M5K  Downtown Toronto   
61        M5L  Downtown Toronto   
66        M5S  Downtown Toronto   
67        M5T  Downtown Toronto   
68        M5V  Downtown Toronto   
69        M5W  Downtown Toronto   
70        M5X  Downtown Toronto   
75        M6G  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  
50                                           Rosedale  43.679563 -79.377529  
51                        Cabbagetown, St. James Town  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Harbourfront, Regent Park  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, King, Adelaide  43.650571 -79.384568  
59  Union Station, Harbourfront East, Toronto Islands  43.640816 -79.381752  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Victoria Hotel, Commerce Court  43.648198 -79.379817  
66                     University of Toronto, Harbord  43.662696 -79.400049  
67          Grange Park, Chinatown, Kensington Market  43.653206 -79.400049  
68  Harbourfront West, CN Tower, Bathurst Quay, Ra...  43.628947 -79.394420  
69                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
70             Underground city, First Canadian Place  43.648429 -79.382280  
75                                           Christie  43.669542 -79.422564

Let's choose the first areas in the table within borough Downtown Toronto, that is postal code M4W, M4X, M4Y at row 50, 51, 52
And see what the results are for this area with foursquare

In [222]:
CP_lat = TOR_NEIGHB.loc[52, 'Latitude'] # postal code latitude value
CP_lon = TOR_NEIGHB.loc[52, 'Longitude'] # postal code longitude value

CP_name = TOR_NEIGHB.loc[52, 'PostalCode'] # Postal Code name
Borough_name = TOR_NEIGHB.loc[52, 'Borough'] # this Postal Code belongs to this Borough

print('Within the borough {}, latitude and longitude values of postal code {} are {}, {}.'.format(Borough_name,
                                                                                                  CP_name,
                                                                                                  CP_lat,
                                                                                                  CP_lon))

Within the borough Downtown Toronto, latitude and longitude values of postal code M4Y are 43.6658599, -79.38315990000001.


In [223]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    CP_lat, 
    CP_lon, 
    radius, 
    LIMIT)
# url # display URL

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c540262db04f57ab2f4b514'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-51a11515498ef97667980b8c-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/arts_entertainment/performingarts_dancestudio_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d134941735',
         'name': 'Dance Studio',
         'pluralName': 'Dance Studios',
         'primary': True,
         'shortName': 'Dance Studio'}],
       'id': '51a11515498ef97667980b8c',
       'location': {'address': '619 Yonge Street',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'btwn Isabella St & Gloucester St',
        'distance': 210,
        'formattedAddress': ['619 Yonge Street (btwn Isabella St & Gloucester St)',

It is interesting to see that while we have a small total of results with M4W (only 4) we have much more results with M4X (46) and M4Y (86). Above the results for M4Y is shown.
We decide to keep a radius of 500 and explore further. As the objective is to find similarities and unsimilarities between areas, we need to see if too much disparity in the numbers of venues may be an issue for relevancy of the data set.
So we will check how many venues we get from requests for each area. An area (postal code) with a very small number of venues may be excluded for the clustering phase. 

So far, we just looked at the number of results we got but let's explore deeper and extract more info. 

In [224]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [225]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name        categories        lat        lng
0   DanceLifeX Centre      Dance Studio  43.666956 -79.385297
1               Smith    Breakfast Spot  43.666927 -79.381421
2  Sansotei Ramen 三草亭  Ramen Restaurant  43.666735 -79.385353
3           The Alley   Bubble Tea Shop  43.665922 -79.385567
4   Glad Day Bookshop         Bookstore  43.665271 -79.380785

Let's confirm wht we briefly saw before before cleaning the json.

In [226]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

86 venues were returned by Foursquare.


Now let's explore for all the areas.
As we started to say above, areas with a very small number of venues will be removed. Let's leverage the function getNearbyVenues from the previous lab And let's add a parameter min_venues to that function that will help determine if we keep each code postal or not. If there are enough venues in this postal code - that is >= min_venues - we keep it and we append the venues to the global list. If not, we don't append the venues for this area in the global list and we won't consider this area in the future map (and clustering).    

In [227]:
# Create a function to repeat a similar process for all the areas (postal codes)

def getNearbyVenues(postalcodes, latitudes, longitudes, radius=500, min_venues=10):
    
    venues_list=[]
# name is actually postal code name
    for name, lat, lng in zip(postalcodes, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        print("results . shape = ", len(results))
        
        # return only relevant information for each nearby venue
        ## if len(results) >= min_venues)
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results if len(results) >= min_venues])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['PostalCode', 
                  'PC Latitude', 
                  'PC Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now we can run the above function on each area and create a new dataframe called TOR_AREAS_VENUES.

In [228]:
TOR_AREAS_VENUES = getNearbyVenues(postalcodes=TOR_NEIGHB['PostalCode'],
                                   latitudes=TOR_NEIGHB['Latitude'],
                                   longitudes=TOR_NEIGHB['Longitude'],
                                  min_venues=10)

M1B
results . shape =  1
M1C
results . shape =  2
M1E
results . shape =  6
M1G
results . shape =  3
M1H
results . shape =  7
M1J
results . shape =  1
M1K
results . shape =  7
M1L
results . shape =  11
M1M
results . shape =  4
M1N
results . shape =  4
M1P
results . shape =  7
M1R
results . shape =  7
M1S
results . shape =  4
M1T
results . shape =  10
M1V
results . shape =  3
M1W
results . shape =  12
M1X
results . shape =  0
M2H
results . shape =  5
M2J
results . shape =  60
M2K
results . shape =  4
M2L
results . shape =  1
M2M
results . shape =  0
M2N
results . shape =  34
M2P
results . shape =  2
M2R
results . shape =  4
M3A
results . shape =  3
M3B
results . shape =  5
M3C
results . shape =  21
M3H
results . shape =  18
M3J
results . shape =  5
M3K
results . shape =  3
M3L
results . shape =  4
M3M
results . shape =  3
M3N
results . shape =  5
M4A
results . shape =  5
M4B
results . shape =  13
M4C
results . shape =  8
M4E
results . shape =  4
M4G
results . shape =  32
M4H
results . sh

With the results above, we can actually see the postal codes that were removed, as the number of venues was lower than the min_venues parameter.
We checked with a 

In [229]:
TOR_AREAS_VENUES.shape

(2012, 7)

Let's check how many venues were returned for each area (pc)

In [230]:
TOR_AREAS_VENUES.groupby('PostalCode').count()

PC Latitude  PC Longitude  Venue  Venue Latitude  Venue Longitude  \
PostalCode                                                                      
M1L                  11            11     11              11               11   
M1T                  10            10     10              10               10   
M1W                  12            12     12              12               12   
M2J                  60            60     60              60               60   
M2N                  34            34     34              34               34   
M3C                  21            21     21              21               21   
M3H                  18            18     18              18               18   
M4B                  13            13     13              13               13   
M4G                  32            32     32              32               32   
M4H                  15            15     15              15               15   
M4K                  42            42     42              42               42   
M4L                  19            19     19              19               19   
M4M                  40            40     40              40               40   
M4R                  19            19     19              19               19   
M4S                  36            36     36              36               36   
M4V                  15            15     15              15               15   
M4X                  46            46     46              46               46   
M4Y                  86            86     86              86               86   
M5A                  48            48     48              48               48   
M5B                 100           100    100             100              100   
M5C                 100           100    100             100              100   
M5E                  55            55     55              55               55   
M5G                  82            82     82              82               82   
M5H                 100           100    100             100              100   
M5J                 100           100    100             100              100   
M5K                 100           100    100             100              100   
M5L                 100           100    100             100              100   
M5M                  23            23     23              23               23   
M5R                  25            25     25              25               25   
M5S                  35            35     35              35               35   
M5T                 100           100    100             100              100   
M5V                  13            13     13              13               13   
M5W                  95            95     95              95               95   
M5X                 100           100    100             100              100   
M6A                  12            12     12              12               12   
M6G                  16            16     16              16               16   
M6H                  21            21     21              21               21   
M6J                  65            65     65              65               65   
M6K                  19            19     19              19               19   
M6P                  23            23     23              23               23   
M6R                  16            16     16              16               16   
M6S                  36            36     36              36               36   
M7A                  44            44     44              44               44   
M7R                  11            11     11              11               11   
M7Y                  18            18     18              18               18   
M8V                  15            15     15              15               15   
M8Z                  11            11     11              11               11   

            Venue Category  
PostalCode   

In [231]:
TOR_AREAS_VENUES.groupby('PostalCode').count().shape

(47, 6)

Trying with several values of min_values, and running :  TOR_AREAS_VENUES.groupby('PostalCode').count().shape
we can tune our filtering of our areas. 
If we set the min number of venues to 10, we still have 47 areas to consider, and 2012 venues globally. 
If we set the min number of venues to 15, we still have 39 areas to consider, and 1919 venues globally. 
If we set the min number of venues to 20, we still have 29 areas to consider, and 1749 venues globally. 


#### Let's find out how many unique categories can be curated from all the returned venues

In [232]:
print('There are {} unique categories.'.format(len(TOR_AREAS_VENUES['Venue Category'].unique())))

There are 256 unique categories.


In [233]:
print('There are {} boroughs.'.format(len(TOR_NEIGHB['Borough'].unique())))
print('There are {} unique areas (postal codes).'.format(len(TOR_AREAS_VENUES['PostalCode'].unique())))

There are 11 boroughs.
There are 47 unique areas (postal codes).


In [234]:
TOR_AREAS_VENUES.head(40)

PostalCode  PC Latitude  PC Longitude  \
0         M1L    43.711112    -79.284577   
1         M1L    43.711112    -79.284577   
2         M1L    43.711112    -79.284577   
3         M1L    43.711112    -79.284577   
4         M1L    43.711112    -79.284577   
5         M1L    43.711112    -79.284577   
6         M1L    43.711112    -79.284577   
7         M1L    43.711112    -79.284577   
8         M1L    43.711112    -79.284577   
9         M1L    43.711112    -79.284577   
10        M1L    43.711112    -79.284577   
11        M1T    43.781638    -79.304302   
12        M1T    43.781638    -79.304302   
13        M1T    43.781638    -79.304302   
14        M1T    43.781638    -79.304302   
15        M1T    43.781638    -79.304302   
16        M1T    43.781638    -79.304302   
17        M1T    43.781638    -79.304302   
18        M1T    43.781638    -79.304302   
19        M1T    43.781638    -79.304302   
20        M1T    43.781638    -79.304302   
21        M1W    43.799525    -79.318389   
22        M1W    43.799525    -79.318389   
23        M1W    43.799525    -79.318389   
24        M1W    43.799525    -79.318389   
25        M1W    43.799525    -79.318389   
26        M1W    43.799525    -79.318389   
27        M1W    43.799525    -79.318389   
28        M1W    43.799525    -79.318389   
29        M1W    43.799525    -79.318389   
30        M1W    43.799525    -79.318389   
31        M1W    43.799525    -79.318389   
32        M1W    43.799525    -79.318389   
33        M2J    43.778517    -79.346556   
34        M2J    43.778517    -79.346556   
35        M2J    43.778517    -79.346556   
36        M2J    43.778517    -79.346556   
37        M2J    43.778517    -79.346556   
38        M2J    43.778517    -79.346556   
39        M2J    43.778517    -79.346556   

                                             Venue  Venue Latitude  \
0                                      Dairy Queen       43.710378   
1                     Warden Ave & St. Clair Ave E       43.712057   
2                               TTC Bus #68 Warden       43.711778   
3                            Warden Subway Station       43.711229   
4                         TTC Bus 102 Markham Road       43.711381   
5                          Warden Station Bus Loop       43.711241   
6                                 Bakery On The Go       43.711271   
7                                   Cafe on the go       43.711151   
8                                Warden Woods Park       43.710527   
9   Edward Jones - Financial Advisor: Thomas Moxam       43.714543   
10                          Clairlea Futbol Centre       43.715234   
11                          Remezzo Italian Bistro       43.778649   
12              The Royal Chinese Restaurant 避風塘小炒       43.780505   
13                                   Eight Noodles       43.778234   
14                                        Kub Khao       43.780438   
15                                  Little Caesars       43.780563   
16                                         Popeyes       43.780476   
17                                             KFC       43.779440   
18                           Enterprise Rent-A-Car       43.779370   
19                                     Gusto Pizza       43.783607   
20                                          Rexall       43.780732   
21                   Mr Congee Chinese Cuisine 龍粥記       43.798879   
22                         Phoenix Restaurant 金鳳餐廳       43.798198   
23                                   Price Chopper       43.799445   
24                                          Subway       43.798983   
25                                     Tim Hortons       43.798281   
26                              Shoppers Drug Mart       43.799670   
27                                             KFC       43.798938   
28                  Yamamoto Japanese Cuisine 山本盛世       43.798589   
29                                      McDonald's       43.798880   
30                          

## Analyze Each Area

In [235]:
# one hot encoding
Toronto_onehot = pd.get_dummies(TOR_AREAS_VENUES[['Venue Category']], prefix="", prefix_sep="")

# add postalcode column back to dataframe
Toronto_onehot['PostalCode'] = TOR_AREAS_VENUES['PostalCode'] 

# move postalcode column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

PostalCode  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0        M1L                  0               0                  0        0   
1        M1L                  0               0                  0        0   
2        M1L                  0               0                  0        0   
3        M1L                  0               0                  0        0   
4        M1L                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0            0           0                    0   
1                 0            0           0                    0   
2                 0            0           0                    0   
3                 0            0           0                    0   
4                 0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Basketball Stadium  Beach  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0      0         0           0                   0   
1                   0      0         0           0                   0   
2                   0      0         0           0                   0   
3                   0      0         0           0                   0   
4                   0      0         0           0                   0   

   Bike Shop  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0          0       0           0              0          0         0   
1          0       0           0              0          0         0   
2          0       0           0              0          0         0   
3          0       0           0              0          0         0   
4          0       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0    

The new dataframe size is :

In [236]:
Toronto_onehot.shape

(2012, 257)

#### Next, let's group rows by postal code and by taking the mean of the frequency of occurrence of each category

In [237]:
Toronto_grouped = Toronto_onehot.groupby('PostalCode').mean().reset_index()
Toronto_grouped

PostalCode  Accessories Store  Adult Boutique  Afghan Restaurant   Airport  \
0         M1L           0.000000        0.000000           0.000000  0.000000   
1         M1T           0.000000        0.000000           0.000000  0.000000   
2         M1W           0.000000        0.000000           0.000000  0.000000   
3         M2J           0.000000        0.000000           0.000000  0.000000   
4         M2N           0.000000        0.000000           0.000000  0.000000   
5         M3C           0.000000        0.000000           0.000000  0.000000   
6         M3H           0.000000        0.000000           0.000000  0.000000   
7         M4B           0.000000        0.000000           0.000000  0.000000   
8         M4G           0.000000        0.000000           0.000000  0.000000   
9         M4H           0.000000        0.000000           0.000000  0.000000   
10        M4K           0.000000        0.000000           0.000000  0.000000   
11        M4L           0.000000        0.000000           0.000000  0.000000   
12        M4M           0.000000        0.000000           0.000000  0.000000   
13        M4R           0.000000        0.000000           0.000000  0.000000   
14        M4S           0.000000        0.000000           0.000000  0.000000   
15        M4V           0.000000        0.000000           0.000000  0.000000   
16        M4X           0.000000        0.000000           0.000000  0.000000   
17        M4Y           0.000000        0.011628           0.011628  0.000000   
18        M5A           0.000000        0.000000           0.000000  0.000000   
19        M5B           0.000000        0.000000           0.000000  0.000000   
20        M5C           0.000000        0.000000           0.000000  0.000000   
21        M5E           0.000000        0.000000           0.000000  0.000000   
22        M5G           0.000000        0.000000           0.000000  0.000000   
23        M5H           0.000000        0.000000           0.000000  0.000000   
24        M5J           0.000000        0.000000           0.000000  0.000000   
25        M5K           0.000000        0.000000           0.000000  0.000000   
26        M5L           0.000000        0.000000           0.000000  0.000000   
27        M5M           0.000000        0.000000           0.000000  0.000000   
28        M5R           0.000000        0.000000           0.000000  0.000000   
29        M5S           0.000000        0.000000           0.000000  0.000000   
30        M5T           0.000000        0.000000           0.000000  0.000000   
31        M5V           0.000000        0.000000           0.000000  0.076923   
32        M5W           0.000000        0.000000           0.000000  0.000000   
33        M5X           0.000000        0.000000           0.000000  0.000000   
34        M6A           0.083333        0.000000           0.000000  0.000000   
35        M6G           0.000000        0.000000           0.000000  0.000000   
36        M6H           0.000000        0.000000           0.000000  0.000000   
37        M6J           0.000000        0.000000           0.000000  0.000000   
38        M6K           0.000000        0.000000           0.000000  0.000000   
39        M6P           0.000000        0.000000           0.000000  0.000000   
40        M6R           0.000000        0.000000           0.000000  0.000000   
41        M6S           0.000000        0.000000           0.000000  0.000000   
42        M7A           0.000000        0.000000           0.000000  0.000000   
43        M7R           0.000000        0.000000           0.000000  0.000000   
44        M7Y           0.000000        0.000000           0.000000  0.000000   
45        M8V           0.000000        0.000000           0.000000  0.000000   
46        M8Z           0.000000        0.000000           0.000000  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000     

In [238]:
# To confirm the new size :
Toronto_grouped.shape

(47, 257)

#### Let's print each area along with the top 5 most common venues

In [239]:
num_top_venues = 5

for area in Toronto_grouped['PostalCode']:
    print("----"+area+"----")
    temp = Toronto_grouped[Toronto_grouped['PostalCode'] == area].T.reset_index()
    temp.columns = ['venue cateory','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----M1L----
   venue cateory  freq
0         Bakery  0.18
1       Bus Line  0.18
2  Metro Station  0.09
3   Intersection  0.09
4   Soccer Field  0.09


----M1T----
         venue cateory  freq
0          Pizza Place   0.2
1   Chinese Restaurant   0.1
2             Pharmacy   0.1
3  Fried Chicken Joint   0.1
4         Noodle House   0.1


----M1W----
          venue cateory  freq
0  Fast Food Restaurant  0.17
1           Coffee Shop  0.17
2    Chinese Restaurant  0.17
3   Japanese Restaurant  0.08
4         Grocery Store  0.08


----M2J----
          venue cateory  freq
0        Clothing Store  0.15
1  Fast Food Restaurant  0.08
2           Coffee Shop  0.07
3         Women's Store  0.05
4      Toy / Game Store  0.03


----M2N----
         venue cateory  freq
0     Ramen Restaurant  0.09
1  Japanese Restaurant  0.06
2       Sandwich Place  0.06
3                 Café  0.06
4          Pizza Place  0.06


----M3C----
      venue cateory  freq
0        Beer Store   0.1
1  Asian Restaurant 

Let's put that into a *pandas* dataframe.
First, let's write a function to sort the venues in descending order.
And then let's create the new dataframe and display the top 10 venues for each neighborhood.

In [240]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [241]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['PostalCode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
postalcode_area_venues_sorted = pd.DataFrame(columns=columns)
postalcode_area_venues_sorted['PostalCode'] = Toronto_grouped['PostalCode']

for ind in np.arange(Toronto_grouped.shape[0]):
    postalcode_area_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

postalcode_area_venues_sorted.head()

PostalCode 1st Most Common Venue 2nd Most Common Venue  \
0        M1L              Bus Line                Bakery   
1        M1T           Pizza Place              Pharmacy   
2        M1W  Fast Food Restaurant    Chinese Restaurant   
3        M2J        Clothing Store  Fast Food Restaurant   
4        M2N      Ramen Restaurant   Japanese Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0  Fast Food Restaurant          Intersection      Business Service   
1    Chinese Restaurant   Rental Car Location       Thai Restaurant   
2           Coffee Shop              Pharmacy        Sandwich Place   
3           Coffee Shop         Women's Store                Bakery   
4        Sandwich Place           Coffee Shop           Pizza Place   

  6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0           Bus Station         Metro Station           Soccer Field   
1    Italian Restaurant          Noodle House   Fast Food Restaurant   
2   Japanese Restaurant        Breakfast Spot            Pizza Place   
3            Food Court      Toy / Game Store             Restaurant   
4            Restaurant                  Café  Indonesian Restaurant   

  9th Most Common Venue       10th Most Common Venue  
0                  Park               General Travel  
1   Fried Chicken Joint             Doner Restaurant  
2         Grocery Store  Eastern European Restaurant  
3         Shopping Mall          Japanese Restaurant  
4                Lounge               Ice Cream Shop

## Cluster Toronto area with k-means


Run *k*-means to cluster the neighborhood into 5 clusters.

In [242]:
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('PostalCode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 1, 1, 0, 1, 1, 0, 1, 1], dtype=int32)

In [243]:
kmeans.labels_

array([0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 3, 0, 0, 1, 0, 3, 3, 0, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 0, 0, 0, 0, 2, 3, 3, 4, 0, 0, 0, 0, 0, 0, 0, 3, 3, 0, 0,
       0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each area.

In [244]:
postalcode_area_venues_sorted.head(11)

PostalCode 1st Most Common Venue 2nd Most Common Venue  \
0         M1L              Bus Line                Bakery   
1         M1T           Pizza Place              Pharmacy   
2         M1W  Fast Food Restaurant    Chinese Restaurant   
3         M2J        Clothing Store  Fast Food Restaurant   
4         M2N      Ramen Restaurant   Japanese Restaurant   
5         M3C           Coffee Shop      Asian Restaurant   
6         M3H           Coffee Shop              Pharmacy   
7         M4B  Fast Food Restaurant           Pizza Place   
8         M4G           Coffee Shop   Sporting Goods Shop   
9         M4H     Indian Restaurant           Yoga Studio   
10        M4K      Greek Restaurant           Coffee Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0   Fast Food Restaurant          Intersection      Business Service   
1     Chinese Restaurant   Rental Car Location       Thai Restaurant   
2            Coffee Shop              Pharmacy        Sandwich Place   
3            Coffee Shop         Women's Store                Bakery   
4         Sandwich Place           Coffee Shop           Pizza Place   
5          Grocery Store            Beer Store                   Gym   
6       Sushi Restaurant    Frozen Yogurt Shop   Fried Chicken Joint   
7               Pharmacy             Gastropub        Breakfast Spot   
8          Grocery Store          Burger Joint          Liquor Store   
9            Supermarket          Liquor Store                  Park   
10        Ice Cream Shop             Bookstore    Italian Restaurant   

   6th Most Common Venue  7th Most Common Venue  8th Most Common Venue  \
0            Bus Station          Metro Station           Soccer Field   
1     Italian Restaurant           Noodle House   Fast Food Restaurant   
2    Japanese Restaurant         Breakfast Spot            Pizza Place   
3             Food Court       Toy / Game Store             Restaurant   
4             Restaurant                   Café  Indonesian Restaurant   
5     Dim Sum Restaurant  General Entertainment              Bike Shop   
6   Fast Food Restaurant                  Diner          Deli / Bodega   
7                   Bank     Rock Climbing Spot           Intersection   
8                   Bank             Sports Bar      Fish & Chips Shop   
9            Coffee Shop                    Gym               Pharmacy   
10        Cosmetics Shop                Brewery        Bubble Tea Shop   

   9th Most Common Venue       10th Most Common Venue  
0                   Park               General Travel  
1    Fried Chicken Joint             Doner Restaurant  
2          Grocery Store  Eastern European Restaurant  
3          Shopping Mall          Japanese Restaurant  
4                 Lounge               Ice Cream Shop  
5         Sandwich Place          Sporting Goods Shop  
6            Pizza Place           Chinese Restaurant  
7     Athletics & Sports                    Pet Store  
8             Beer Store                Smoothie Shop  
9            Pizza Place                 Burger Joint  
10            Restaurant                          Pub

In [245]:
postalcode_area_venues_sorted.shape

(47, 11)

In [246]:
# add clustering labels
#### run the following line only once!   
# postalcode_area_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = TOR_NEIGHB

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(postalcode_area_venues_sorted.set_index('PostalCode'), on='PostalCode', how='inner')

# need 'inner' in the join bc we removed some areas (postal codes) before clustering - as they had less than 10 venues. 
# They would come back in the df with the join if not 'inner' option. And they would have NaN in Cluster Labels and else

Toronto_merged.head() 

PostalCode      Borough                             Neighborhood  \
7         M1L  Scarborough          Golden Mile, Oakridge, Clairlea   
13        M1T  Scarborough  Clarks Corners, Tam O'Shanter, Sullivan   
15        M1W  Scarborough            Steeles West, L'Amoreaux West   
18        M2J   North York             Oriole, Fairview, Henry Farm   
22        M2N   North York                         Willowdale South   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
7   43.711112 -79.284577               0              Bus Line   
13  43.781638 -79.304302               0           Pizza Place   
15  43.799525 -79.318389               1  Fast Food Restaurant   
18  43.778517 -79.346556               1        Clothing Store   
22  43.770120 -79.408493               0      Ramen Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7                 Bakery  Fast Food Restaurant          Intersection   
13              Pharmacy    Chinese Restaurant   Rental Car Location   
15    Chinese Restaurant           Coffee Shop              Pharmacy   
18  Fast Food Restaurant           Coffee Shop         Women's Store   
22   Japanese Restaurant        Sandwich Place           Coffee Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
7       Business Service           Bus Station         Metro Station   
13       Thai Restaurant    Italian Restaurant          Noodle House   
15        Sandwich Place   Japanese Restaurant        Breakfast Spot   
18                Bakery            Food Court      Toy / Game Store   
22           Pizza Place            Restaurant                  Café   

    8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
7            Soccer Field                  Park               General Travel  
13   Fast Food Restaurant   Fried Chicken Joint             Doner Restaurant  
15            Pizza Place         Grocery Store  Eastern European Restaurant  
18             Restaurant         Shopping Mall          Japanese Restaurant  
22  Indonesian Restaurant                Lounge               Ice Cream Shop

In [247]:
Toronto_merged

PostalCode           Borough  \
7         M1L       Scarborough   
13        M1T       Scarborough   
15        M1W       Scarborough   
18        M2J        North York   
22        M2N        North York   
27        M3C        North York   
28        M3H        North York   
35        M4B         East York   
38        M4G         East York   
39        M4H         East York   
41        M4K      East Toronto   
42        M4L      East Toronto   
43        M4M      East Toronto   
46        M4R   Central Toronto   
47        M4S   Central Toronto   
49        M4V   Central Toronto   
51        M4X  Downtown Toronto   
52        M4Y  Downtown Toronto   
53        M5A  Downtown Toronto   
54        M5B  Downtown Toronto   
55        M5C  Downtown Toronto   
56        M5E  Downtown Toronto   
57        M5G  Downtown Toronto   
58        M5H  Downtown Toronto   
59        M5J  Downtown Toronto   
60        M5K  Downtown Toronto   
61        M5L  Downtown Toronto   
62        M5M        North York   
65        M5R   Central Toronto   
66        M5S  Downtown Toronto   
67        M5T  Downtown Toronto   
68        M5V  Downtown Toronto   
69        M5W  Downtown Toronto   
70        M5X  Downtown Toronto   
71        M6A        North York   
75        M6G  Downtown Toronto   
76        M6H      West Toronto   
77        M6J      West Toronto   
78        M6K      West Toronto   
82        M6P      West Toronto   
83        M6R      West Toronto   
84        M6S      West Toronto   
85        M7A      Queen's Park   
86        M7R       Mississauga   
87        M7Y      East Toronto   
88        M8V         Etobicoke   
92        M8Z         Etobicoke   

                                         Neighborhood   Latitude  Longitude  \
7                     Golden Mile, Oakridge, Clairlea  43.711112 -79.284577   
13            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302   
15                      Steeles West, L'Amoreaux West  43.799525 -79.318389   
18                       Oriole, Fairview, Henry Farm  43.778517 -79.346556   
22                                   Willowdale South  43.770120 -79.408493   
27                   Don Mills South, Flemingdon Park  43.725900 -79.340923   
28    Downsview North, Wilson Heights, Bathurst Manor  43.754328 -79.442259   
35                    Parkview Hill, Woodbine Gardens  43.706397 -79.309937   
38                                            Leaside  43.709060 -79.363452   
39                                   Thorncliffe Park  43.705369 -79.349372   
41                       The Danforth West, Riverdale  43.679557 -79.352188   
42                     India Bazaar, The Beaches West  43.668999 -79.315572   
43                                    Studio District  43.659526 -79.340923   
46                                 North Toronto West  43.715383 -79.405678   
47                                         Davisville  43.704324 -79.388790   
49  Rathnelly, South Hill, Summerhill West, Forest...  43.686412 -79.400049   
51                        Cabbagetown, St. James Town  43.667967 -79.367675   
52                               Church and Wellesley  43.665860 -79.383160   
53                          Harbourfront, Regent Park  43.654260 -79.360636   
54                           Garden District, Ryerson  43.657162 -79.378937   
55                                     St. James Town  43.651494 -79.375418   
56                                        Berczy Park  43.644771 -79.373306   
57                                 Central Bay Street  43.657952 -79.387383   
58                           Richmond, King, Adelaide  43.650571 -79.384568   
59  Union Station, Harbourfront East, Toronto Islands  43.640816 -79.381752   
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
61                     Victoria Hotel, Commerce Court  43.648198 -79.379817   
62                  Bedford Park, Lawrence Manor East  43.733283 -79.419750   
65                The Annex, Yo

In [248]:
Toronto_merged.shape

(47, 16)

Finally, let's visualize the resulting clusters

In [249]:
latitude

Latitude
50  43.679563

In [250]:
# create map
map_clusters = folium.Map(location=[lat, lon], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# Toronto_merged['Borough']

# add markers to the map
markers_colors = []
for lat, lon, borough, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Borough'], Toronto_merged['PostalCode'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(borough) + ', ' + str(poi) + ', Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

##  Examine Clusters

Now, we can examine each cluster and determine the discriminating venue categories that distinguish each cluster.

The 5 clusters are listed below

Cluster 0

In [261]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[list(range(0,3)) + list(range(5, Toronto_merged.shape[1]))]]

PostalCode           Borough  \
7         M1L       Scarborough   
13        M1T       Scarborough   
22        M2N        North York   
35        M4B         East York   
42        M4L      East Toronto   
43        M4M      East Toronto   
47        M4S   Central Toronto   
52        M4Y  Downtown Toronto   
62        M5M        North York   
65        M5R   Central Toronto   
66        M5S  Downtown Toronto   
67        M5T  Downtown Toronto   
75        M6G  Downtown Toronto   
76        M6H      West Toronto   
77        M6J      West Toronto   
78        M6K      West Toronto   
82        M6P      West Toronto   
83        M6R      West Toronto   
84        M6S      West Toronto   
87        M7Y      East Toronto   
88        M8V         Etobicoke   
92        M8Z         Etobicoke   

                                         Neighborhood  Cluster Labels  \
7                     Golden Mile, Oakridge, Clairlea               0   
13            Clarks Corners, Tam O'Shanter, Sullivan               0   
22                                   Willowdale South               0   
35                    Parkview Hill, Woodbine Gardens               0   
42                     India Bazaar, The Beaches West               0   
43                                    Studio District               0   
47                                         Davisville               0   
52                               Church and Wellesley               0   
62                  Bedford Park, Lawrence Manor East               0   
65                The Annex, Yorkville, North Midtown               0   
66                     University of Toronto, Harbord               0   
67          Grange Park, Chinatown, Kensington Market               0   
75                                           Christie               0   
76                       Dovercourt Village, Dufferin               0   
77                           Little Portugal, Trinity               0   
78       Exhibition Place, Brockton, Parkdale Village               0   
82                      The Junction South, High Park               0   
83                             Parkdale, Roncesvalles               0   
84                                 Runnymede, Swansea               0   
87  Business Reply Mail Processing Centre 969 Eastern               0   
88       Humber Bay Shores, Mimico South, New Toronto               0   
92  Royal York South West, South of Bloor, The Que...               0   

   1st Most Common Venue 2nd Most Common Venue          3rd Most Common Venue  \
7               Bus Line                Bakery           Fast Food Restaurant   
13           Pizza Place              Pharmacy             Chinese Restaurant   
22      Ramen Restaurant   Japanese Restaurant                 Sandwich Place   
35  Fast Food Restaurant           Pizza Place                       Pharmacy   
42                  Park  Fast Food Restaurant                    Pizza Place   
43                  Café           Coffee Shop             Italian Restaurant   
47           Pizza Place        Sandwich Place                   Dessert Shop   
52   Japanese Restaurant           Coffee Shop               Sushi Restaurant   
62  Fast Food Restaurant           Coffee Shop             Italian Restaurant   
65                  Café           Coffee Shop                 Sandwich Place   
66                  Café            Restaurant                            Bar   
67                  Café                   Bar  Vegetarian / Vegan Restaurant   
75         Grocery Store                  Café                           Park   
76           Supermarket                Bakery                       Pharmacy   
77                   Bar           Men's Store                           Café   
78                  Café        Breakfast Spot                    Coffee Shop   
82                  Café    Mexican Restaurant                  Grocery Store   
83        Breakfast Spot             Gift Shop             

Cluster 1

In [262]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[list(range(0,3)) + list(range(5, Toronto_merged.shape[1]))]]

PostalCode          Borough  \
15        M1W      Scarborough   
18        M2J       North York   
27        M3C       North York   
28        M3H       North York   
38        M4G        East York   
39        M4H        East York   
46        M4R  Central Toronto   

                                       Neighborhood  Cluster Labels  \
15                    Steeles West, L'Amoreaux West               1   
18                     Oriole, Fairview, Henry Farm               1   
27                 Don Mills South, Flemingdon Park               1   
28  Downsview North, Wilson Heights, Bathurst Manor               1   
38                                          Leaside               1   
39                                 Thorncliffe Park               1   
46                               North Toronto West               1   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
15  Fast Food Restaurant    Chinese Restaurant           Coffee Shop   
18        Clothing Store  Fast Food Restaurant           Coffee Shop   
27           Coffee Shop      Asian Restaurant         Grocery Store   
28           Coffee Shop              Pharmacy      Sushi Restaurant   
38           Coffee Shop   Sporting Goods Shop         Grocery Store   
39     Indian Restaurant           Yoga Studio           Supermarket   
46   Sporting Goods Shop           Coffee Shop           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
15              Pharmacy        Sandwich Place   Japanese Restaurant   
18         Women's Store                Bakery            Food Court   
27            Beer Store                   Gym    Dim Sum Restaurant   
28    Frozen Yogurt Shop   Fried Chicken Joint  Fast Food Restaurant   
38          Burger Joint          Liquor Store                  Bank   
39          Liquor Store                  Park           Coffee Shop   
46  Gym / Fitness Center  Fast Food Restaurant                 Diner   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
15         Breakfast Spot           Pizza Place         Grocery Store   
18       Toy / Game Store            Restaurant         Shopping Mall   
27  General Entertainment             Bike Shop        Sandwich Place   
28                  Diner         Deli / Bodega           Pizza Place   
38             Sports Bar     Fish & Chips Shop            Beer Store   
39                    Gym              Pharmacy           Pizza Place   
46           Dessert Shop    Mexican Restaurant    Miscellaneous Shop   

         10th Most Common Venue  
15  Eastern European Restaurant  
18          Japanese Restaurant  
27          Sporting Goods Shop  
28           Chinese Restaurant  
38                Smoothie Shop  
39                 Burger Joint  
46                         Park

Cluster 2

In [263]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[list(range(0,3)) + list(range(5, Toronto_merged.shape[1]))]]

PostalCode           Borough  \
68        M5V  Downtown Toronto   

                                         Neighborhood  Cluster Labels  \
68  Harbourfront West, CN Tower, Bathurst Quay, Ra...               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
68        Airport Lounge      Airport Terminal       Airport Service   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
68       Harbor / Marina      Sculpture Garden              Boutique   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
68         Boat or Ferry          Airport Gate               Airport   

   10th Most Common Venue  
68     Airport Food Court

Cluster 3

In [264]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[list(range(0,3)) + list(range(5, Toronto_merged.shape[1]))]]

PostalCode           Borough  \
41        M4K      East Toronto   
49        M4V   Central Toronto   
51        M4X  Downtown Toronto   
53        M5A  Downtown Toronto   
54        M5B  Downtown Toronto   
55        M5C  Downtown Toronto   
56        M5E  Downtown Toronto   
57        M5G  Downtown Toronto   
58        M5H  Downtown Toronto   
59        M5J  Downtown Toronto   
60        M5K  Downtown Toronto   
61        M5L  Downtown Toronto   
69        M5W  Downtown Toronto   
70        M5X  Downtown Toronto   
85        M7A      Queen's Park   
86        M7R       Mississauga   

                                         Neighborhood  Cluster Labels  \
41                       The Danforth West, Riverdale               3   
49  Rathnelly, South Hill, Summerhill West, Forest...               3   
51                        Cabbagetown, St. James Town               3   
53                          Harbourfront, Regent Park               3   
54                           Garden District, Ryerson               3   
55                                     St. James Town               3   
56                                        Berczy Park               3   
57                                 Central Bay Street               3   
58                           Richmond, King, Adelaide               3   
59  Union Station, Harbourfront East, Toronto Islands               3   
60           Toronto Dominion Centre, Design Exchange               3   
61                     Victoria Hotel, Commerce Court               3   
69                    Stn A PO Boxes 25 The Esplanade               3   
70             Underground city, First Canadian Place               3   
85                                       Queen's Park               3   
86              Canada Post Gateway Processing Centre               3   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
41      Greek Restaurant           Coffee Shop        Ice Cream Shop   
49           Coffee Shop                   Pub   Fried Chicken Joint   
51           Coffee Shop            Restaurant                Bakery   
53           Coffee Shop                  Park                  Café   
54           Coffee Shop        Clothing Store                  Café   
55           Coffee Shop            Restaurant                 Hotel   
56           Coffee Shop            Restaurant          Cocktail Bar   
57           Coffee Shop                  Café    Italian Restaurant   
58           Coffee Shop                  Café       Thai Restaurant   
59           Coffee Shop              Aquarium                 Hotel   
60           Coffee Shop                 Hotel                  Café   
61           Coffee Shop                  Café            Restaurant   
69           Coffee Shop            Restaurant                  Café   
70                  Café           Coffee Shop                 Hotel   
85           Coffee Shop   Japanese Restaurant      Sushi Restaurant   
86           Coffee Shop                 Hotel  Gym / Fitness Center   

        4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
41                  Bookstore    Italian Restaurant        Cosmetics Shop   
49                Pizza Place           Supermarket            Bagel Shop   
51         Italian Restaurant                   Pub                  Park   
53                        Pub                Bakery        Breakfast Spot   
54  Middle Eastern Restaurant        Cosmetics Shop       Thai Restaurant   
55                       Café        Clothing Store                  Park   
56                       Café            Steakhouse              Beer Bar   
57                        Bar          Burger Joint        Sandwich Place   
58        American Restaurant            Steakhouse                   Gym   
59                Pizza Place                  Café               Brewery   
60        American Restaurant            Restaurant    Italian Restaurant   
61              

Cluster 4

In [266]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[list(range(0,3)) + list(range(5, Toronto_merged.shape[1]))]]

PostalCode     Borough                      Neighborhood  Cluster Labels  \
71        M6A  North York  Lawrence Heights, Lawrence Manor               4   

   1st Most Common Venue 2nd Most Common Venue  3rd Most Common Venue  \
71        Clothing Store     Accessories Store  Vietnamese Restaurant   

   4th Most Common Venue 5th Most Common Venue   6th Most Common Venue  \
71    Miscellaneous Shop              Boutique  Furniture / Home Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
71           Coffee Shop           Event Space  Ethiopian Restaurant   

   10th Most Common Venue  
71      Electronics Store

Thank you for reviewing !
Your feedbacks are welcome.